### PR TITLE
Added support for CRL check

### PR DESCRIPTION
**Pull request by tsaarni**
This change adds support for CRL (Certificate Revocation List) check, to validate that the certificate of the syslog server has not been revoked. The CRL is loaded from file, meaning it must be pre-loaded on the system where logstash is executing.

----

Part of https://github.com/elastic/logstash/issues/15236

### CODE REVIEW

**Comment by tsaarni** [comment]
Tests do not work anymore with current logstash version because of #51.

File: lib/logstash/outputs/syslog.rb [hunk 1]
```diff
@@ -223,6 +229,8 @@ def connect
     socket
   end
 
+  CRL_END_TAG = "\n-----END X509 CRL-----\n"
```
Comment (andsel) [inline lib/logstash/outputs/syslog.rb:232]: This constant definition has to be moved on top of the file, close to the *_LABEL constants

File: lib/logstash/outputs/syslog.rb [hunk 2]
```diff
@@ -237,6 +245,14 @@ def setup_ssl
       else
         cert_store.add_file(@ssl_cacert)
       end
+      if @ssl_crl
+        # copy the behavior of X509_load_crl_file() which supports loading bundles of CRLs.
+        File.read(@ssl_crl).split(CRL_END_TAG).each do |crl|
+          crl << CRL_END_TAG
+          cert_store.add_crl(OpenSSL::X509::CRL.new(crl))
```
Comment (andsel) [inline lib/logstash/outputs/syslog.rb:252]: What happens if the CRL file is malformed or contains malformed certificates?

**Review by andsel — CHANGES_REQUESTED** [review]
I have some doubts on how the code behaves if the file is malformed. Maybe those error could be catched and shape a better ConfigurationError exception.
How could I test this?
Could you cover the feature with some unit test?

File: lib/logstash/outputs/syslog.rb (hunk 1 above)
Comment (tsaarni) [reply lib/logstash/outputs/syslog.rb]: Done!

File: lib/logstash/outputs/syslog.rb (hunk 2 above)
Comment (tsaarni) [reply lib/logstash/outputs/syslog.rb:263]: It seems to behave similarly as for other malformed .pem files: register will fail.  In this case, register will fail with `OpenSSL::X509::CRLError` with error description "malformed PEM data".

**Comment by tsaarni** [comment]
Sorry for delay! 

The plugin is lacking _all_ TLS related tests, so I made an attempt to add some, including tests for reading malformed PEM files. I took inspiration from https://github.com/logstash-plugins/logstash-output-tcp.

I needed some test certificates, which I committed in the repo. I used my own tool [certyaml](https://github.com/tsaarni/certyaml) to create those, I hope that is OK.

> I have some doubts on how the code behaves if the file is malformed. Maybe those error could be catched and shape a better ConfigurationError exception. How could I test this? Could you cover the feature with some unit test?

It seems to fail in a same way as it has been failing if user would pass invalid certificate or private key. For some reason, there is no error when loading malformed CA certificate, which may be a problem jruby-openssl.

File: spec/outputs/syslog_tls_spec.rb [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    #     "ssl_key" => File.join(FIXTURES_PATH, "client-key.pem"),
+    #     "ssl_cacert" => File.join(FIXTURES_PATH, "invalid.pem"),
+    #     "ssl_crl"  => File.join(FIXTURES_PATH, "ca-crl.pem")
+    #   ) }
+
+    #   it "register raises error" do
+    #     expect { subject.register }.to raise_error(OpenSSL::PKey::RSAError, /malformed PEM data/)
+    #   end
+    # end
+
```
Comment (andsel) [inline spec/outputs/syslog_tls_spec.rb:150]:
```suggestion
```

Remove commented code

File: spec/outputs/syslog_tls_spec.rb [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      let(:options ) { super().merge("ssl_verify" => true) }
+      let(:server_cert_file) { File.join(FIXTURES_PATH, "untrusted-server.pem") }
+      let(:server_pkey_file) { File.join(FIXTURES_PATH, "untrusted-server-key.pem") }
+
+      it "syslog output refuses to connect" do
+        Thread.start { secure_server.accept rescue nil }
+        expect(subject.logger).to receive(:error).with(/SSL Error/i, hash_including(exception: OpenSSL::SSL::SSLError)).once.and_throw :TEST_DONE
+        expect { subject.receive event }.to throw_symbol(:TEST_DONE)
+      end
+    end
```
Comment (andsel) [inline spec/outputs/syslog_tls_spec.rb:92]:
```suggestion
    context "server with untrusted certificates" do
      let(:server_cert_file) { File.join(FIXTURES_PATH, "untrusted-server.pem") }
      let(:server_pkey_file) { File.join(FIXTURES_PATH, "untrusted-server-key.pem") }

      context "ss_verify disabled" do
        let(:options ) { super().merge("ssl_verify" => false) }

        it_behaves_like "syslog output"
      end      

      context "ss_verify enabled" do
        let(:options ) { super().merge("ssl_verify" => true) }

        it "should refuse to connect" do
          Thread.start { secure_server.accept rescue nil }
          expect(subject.logger).to receive(:error).with(/SSL Error/i, hash_including(exception: OpenSSL::SSL::SSLError)).once.and_throw :TEST_DONE
          expect { subject.receive event }.to throw_symbol(:TEST_DONE)
        end
      end
    end
```

A nested context could make it even more readable

**Review by andsel — CHANGES_REQUESTED** [review]
A great thank you @tsaarni  for huge amount of work on this!

I think we are pretty close, left a couple of suggestions.

File: spec/outputs/syslog_tls_spec.rb (hunk 3 above)
Comment (tsaarni) [reply spec/outputs/syslog_tls_spec.rb]: Removed.

File: spec/outputs/syslog_tls_spec.rb (hunk 4 above)
Comment (tsaarni) [reply spec/outputs/syslog_tls_spec.rb]: True, changed accordingly!

**Comment by tsaarni** [comment]
> A great thank you @tsaarni for huge amount of work on this!
> 
> I think we are pretty close, left a couple of suggestions.

No problem, adding TLS tests is a good learning experience for a ruby newbie like me :)

**Review by andsel — APPROVED** [review]
LGTM

**Review by andsel — CHANGES_REQUESTED** [review]
LGTM, please add a line to the Changelog file describing this feature.

**Review by andsel — APPROVED** [review]
LGTM

**Pull request merged**